Provider produced inconsistent result after apply on aws_iam_role_policy_attachment

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

Terraform v0.13.5
+ provider registry.terraform.io/hashicorp/aws v3.14.1
+ provider registry.terraform.io/hashicorp/kubernetes v1.13.3
+ provider registry.terraform.io/hashicorp/local v2.0.0

### Affected Resource(s)

aws_iam_role_policy_attachment 

### Terraform Configuration Files

```terraform
resource "aws_iam_role" "eks-instances-role" {
  name = "eks-instances-role-${terraform.workspace}"
  force_detach_policies = true

  assume_role_policy = <<EOF
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Action": "sts:AssumeRole",
            "Principal": {
                "Service": "eks.amazonaws.com"
            },
            "Effect": "Allow",
            "Sid": ""
        }
    ]
}
EOF
}

resource "aws_iam_role_policy_attachment" "main-node-AmazonEKSWorkerNodePolicy" {
  policy_arn = "arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy"
  role       = aws_iam_role.eks-instances-role.name
}
```

### Debug Output

```console
2020/11/17 14:18:53 [DEBUG] module.dso-eks.aws_iam_role_policy_attachment.main-node-AmazonEKS_CNI_Policy: apply errored, but we're indicating that via the Error pointer rather than returning it: Provider produced inconsistent result after apply: When applying changes to module.dso-eks.aws_iam_role_policy_attachment.main-node-AmazonEKS_CNI_Policy, provider "registry.terraform.io/hashicorp/aws" produced an unexpected new value: Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own issue tracker.
2020/11/17 14:18:53 [TRACE] eval: *terraform.EvalMaybeTainted
2020/11/17 14:18:53 [TRACE] EvalMaybeTainted: module.dso-eks.aws_iam_role_policy_attachment.main-node-AmazonEKS_CNI_Policy encountered an error during creation, so it is now marked as tainted
2020/11/17 14:18:53 [TRACE] eval: *terraform.EvalWriteState
2020/11/17 14:18:53 [TRACE] EvalWriteState: removing state object for module.dso-eks.aws_iam_role_policy_attachment.main-node-AmazonEKS_CNI_Policy
2020/11/17 14:18:53 [TRACE] eval: *terraform.EvalApplyProvisioners
2020/11/17 14:18:53 [TRACE] EvalApplyProvisioners: aws_iam_role_policy_attachment.main-node-AmazonEKS_CNI_Policy has no state, so skipping provisioners
2020/11/17 14:18:53 [TRACE] eval: *terraform.EvalMaybeTainted
2020/11/17 14:18:53 [TRACE] EvalMaybeTainted: module.dso-eks.aws_iam_role_policy_attachment.main-node-AmazonEKS_CNI_Policy encountered an error during creation, so it is now marked as tainted
2020/11/17 14:18:53 [TRACE] eval: *terraform.EvalWriteState
2020/11/17 14:18:53 [TRACE] EvalWriteState: removing state object for module.dso-eks.aws_iam_role_policy_attachment.main-node-AmazonEKS_CNI_Policy
2020/11/17 14:18:53 [TRACE] eval: *terraform.EvalIf
2020/11/17 14:18:53 [TRACE] eval: *terraform.EvalIf
2020/11/17 14:18:53 [TRACE] eval: *terraform.EvalWriteDiff
2020/11/17 14:18:53 [TRACE] eval: *terraform.EvalApplyPost
2020/11/17 14:18:53 [ERROR] eval: *terraform.EvalApplyPost, err: Provider produced inconsistent result after apply: When applying changes to module.dso-eks.aws_iam_role_policy_attachment.main-node-AmazonEKS_CNI_Policy, provider "registry.terraform.io/hashicorp/aws" produced an unexpected new value: Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own issue tracker.
2020/11/17 14:18:53 [ERROR] eval: *terraform.EvalSequence, err: Provider produced inconsistent result after apply: When applying changes to module.dso-eks.aws_iam_role_policy_attachment.main-node-AmazonEKS_CNI_Policy, provider "registry.terraform.io/hashicorp/aws" produced an unexpected new value: Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own issue tracker.
2020/11/17 14:18:53 [TRACE] [walkApply] Exiting eval tree: module.dso-eks.aws_iam_role_policy_attachment.main-node-AmazonEKS_CNI_Policy
2020/11/17 14:18:53 [TRACE] vertex "module.dso-eks.aws_iam_role_policy_attachment.main-node-AmazonEKS_CNI_Policy": visit complete
```

### Expected Behavior

Policy was attached to role without error

### Actual Behavior

```console
Error: Provider produced inconsistent result after apply

When applying changes to
module.dso-eks.aws_iam_role_policy_attachment.main-node-AmazonEKS_CNI_Policy,
provider "registry.terraform.io/hashicorp/aws" produced an unexpected new
value: Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

--- a/.changelog/18435.txt
+++ b/.changelog/18435.txt
@@ -1,0 +1,11 @@
+```release-note:bug
+resource/aws_iam_role: Handle read-after-create eventual consistency
+```
+
+```release-note:bug
+resource/aws_iam_role_policy: Handle read-after-create eventual consistency
+```
+
+```release-note:bug
+resource/aws_iam_role_policy_attachment: Handle read-after-create eventual consistency
+```

--- a/aws/resource_aws_iam_role.go
+++ b/aws/resource_aws_iam_role.go
@@ -16,6 +16,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
 	"github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags"
 	"github.com/terraform-providers/terraform-provider-aws/aws/internal/service/iam/waiter"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/tfresource"
 )
 
 func resourceAwsIamRole() *schema.Resource {
@@ -235,20 +236,40 @@ func resourceAwsIamRoleRead(d *schema.ResourceData, meta interface{}) error {
 		RoleName: aws.String(d.Id()),
 	}
 
-	getResp, err := iamconn.GetRole(request)
-	if err != nil {
-		if isAWSErr(err, iam.ErrCodeNoSuchEntityException, "") {
-			log.Printf("[WARN] IAM Role %q not found, removing from state", d.Id())
-			d.SetId("")
-			return nil
+	var getResp *iam.GetRoleOutput
+
+	err := resource.Retry(waiter.PropagationTimeout, func() *resource.RetryError {
+		var err error
+
+		getResp, err = iamconn.GetRole(request)
+
+		if d.IsNewResource() && tfawserr.ErrCodeEquals(err, iam.ErrCodeNoSuchEntityException) {
+			return resource.RetryableError(err)
 		}
-		return fmt.Errorf("Error reading IAM Role %s: %s", d.Id(), err)
+
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+
+		return nil
+	})
+
+	if tfresource.TimedOut(err) {
+		getResp, err = iamconn.GetRole(request)
+	}
+
+	if !d.IsNewResource() && tfawserr.ErrCodeEquals(err, iam.ErrCodeNoSuchEntityException) {
+		log.Printf("[WARN] IAM Role (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	if err != nil {
+		return fmt.Errorf("error reading IAM Role (%s): %w", d.Id(), err)
 	}
 
 	if getResp == nil || getResp.Role == nil {
-		log.Printf("[WARN] IAM Role %q not found, removing from state", d.Id())
-		d.SetId("")
-		return nil
+		return fmt.Errorf("error reading IAM Role (%s): empty response", d.Id())
 	}
 
 	role := getResp.Role

--- a/aws/resource_aws_iam_role_policy.go
+++ b/aws/resource_aws_iam_role_policy.go
@@ -8,8 +8,11 @@ import (
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/hashicorp/aws-sdk-go-base/tfawserr"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/service/iam/waiter"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/tfresource"
 )
 
 func resourceAwsIamRolePolicy() *schema.Resource {
@@ -94,18 +97,40 @@ func resourceAwsIamRolePolicyRead(d *schema.ResourceData, meta interface{}) erro
 		RoleName:   aws.String(role),
 	}
 
-	getResp, err := iamconn.GetRolePolicy(request)
-	if err != nil {
-		if isAWSErr(err, iam.ErrCodeNoSuchEntityException, "") {
-			log.Printf("[WARN] IAM Role Policy (%s) for %s not found, removing from state", name, role)
-			d.SetId("")
-			return nil
+	var getResp *iam.GetRolePolicyOutput
+
+	err = resource.Retry(waiter.PropagationTimeout, func() *resource.RetryError {
+		var err error
+
+		getResp, err = iamconn.GetRolePolicy(request)
+
+		if d.IsNewResource() && tfawserr.ErrCodeEquals(err, iam.ErrCodeNoSuchEntityException) {
+			return resource.RetryableError(err)
 		}
-		return fmt.Errorf("Error reading IAM policy %s from role %s: %s", name, role, err)
+
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+
+		return nil
+	})
+
+	if tfresource.TimedOut(err) {
+		getResp, err = iamconn.GetRolePolicy(request)
 	}
 
-	if getResp.PolicyDocument == nil {
-		return fmt.Errorf("GetRolePolicy returned a nil policy document")
+	if !d.IsNewResource() && tfawserr.ErrCodeEquals(err, iam.ErrCodeNoSuchEntityException) {
+		log.Printf("[WARN] IAM Role Policy (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	if err != nil {
+		return fmt.Errorf("error reading IAM Role Policy (%s): %w", d.Id(), err)
+	}
+
+	if getResp == nil || getResp.PolicyDocument == nil {
+		return fmt.Errorf("error reading IAM Role Policy (%s): empty response", d.Id())
 	}
 
 	policy, err := url.QueryUnescape(*getResp.PolicyDocument)

--- a/aws/resource_aws_iam_role_policy_attachment.go
+++ b/aws/resource_aws_iam_role_policy_attachment.go
@@ -7,8 +7,11 @@ import (
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/hashicorp/aws-sdk-go-base/tfawserr"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/service/iam/waiter"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/tfresource"
 )
 
 func resourceAwsIamRolePolicyAttachment() *schema.Resource {
@@ -56,21 +59,49 @@ func resourceAwsIamRolePolicyAttachmentRead(d *schema.ResourceData, meta interfa
 	conn := meta.(*AWSClient).iamconn
 	role := d.Get("role").(string)
 	policyARN := d.Get("policy_arn").(string)
+	// Human friendly ID for error messages since d.Id() is non-descriptive
+	id := fmt.Sprintf("%s:%s", role, policyARN)
 
-	hasPolicyAttachment, err := iamRoleHasPolicyARNAttachment(conn, role, policyARN)
+	var hasPolicyAttachment bool
 
-	if isAWSErr(err, iam.ErrCodeNoSuchEntityException, "") {
-		log.Printf("[WARN] IAM Role (%s) not found, removing from state", role)
+	err := resource.Retry(waiter.PropagationTimeout, func() *resource.RetryError {
+		var err error
+
+		hasPolicyAttachment, err = iamRoleHasPolicyARNAttachment(conn, role, policyARN)
+
+		if d.IsNewResource() && tfawserr.ErrCodeEquals(err, iam.ErrCodeNoSuchEntityException) {
+			return resource.RetryableError(err)
+		}
+
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+
+		if d.IsNewResource() && !hasPolicyAttachment {
+			return resource.RetryableError(&resource.NotFoundError{
+				LastError: fmt.Errorf("IAM Role Managed Policy Attachment (%s) not found", id),
+			})
+		}
+
+		return nil
+	})
+
+	if tfresource.TimedOut(err) {
+		hasPolicyAttachment, err = iamRoleHasPolicyARNAttachment(conn, role, policyARN)
+	}
+
+	if !d.IsNewResource() && tfawserr.ErrCodeEquals(err, iam.ErrCodeNoSuchEntityException) {
+		log.Printf("[WARN] IAM Role Managed Policy Attachment (%s) not found, removing from state", id)
 		d.SetId("")
 		return nil
 	}
 
 	if err != nil {
-		return fmt.Errorf("error finding IAM Role (%s) Policy Attachment (%s): %s", role, policyARN, err)
+		return fmt.Errorf("error reading IAM Role Managed Policy Attachment (%s): %w", id, err)
 	}
 
-	if !hasPolicyAttachment {
-		log.Printf("[WARN] IAM Role (%s) Policy Attachment (%s) not found, removing from state", role, policyARN)
+	if !d.IsNewResource() && !hasPolicyAttachment {
+		log.Printf("[WARN] IAM Role Managed Policy Attachment (%s) not found, removing from state", id)
 		d.SetId("")
 		return nil
 	}

